omnigibson/utils/bddl_utils.py

```diff
@@ -89,9 +127,810 @@ def process_single_condition(condition):
     "open": get_unary_predicate_for_state(object_states.Open, "open"),
     "toggled_on": get_unary_predicate_for_state(object_states.ToggledOn, "toggled_on"),
     "frozen": get_unary_predicate_for_state(object_states.Frozen, "frozen"),
+    "future": ObjectStateFuturePredicate,
+    "real": ObjectStateRealPredicate,
 }
 
+KINEMATIC_STATES_BDDL = frozenset([state.__name__.lower() for state in _KINEMATIC_STATE_SET])
+
+# Load substance and object mapping
+with open(f"{bddl.__path__[0]}/../substance_synset_mapping.json", "r") as f:
+    SUBSTANCE_SYNSET_MAPPING = json.load(f)
+
+
+# BEHAVIOR-related
+OBJECT_TAXONOMY = ObjectTaxonomy() #hierarchy_type="b1k")
+FLOOR_SYNSET = "floor.n.01"
+with open(os.path.join(os.path.dirname(bddl.__file__), "activity_manifest.txt")) as f:
+    BEHAVIOR_ACTIVITIES = {line.strip() for line in f.readlines()}
+NON_SAMPLEABLE_SYNSETS = set()
+non_sampleable_category_txt = os.path.join(gm.DATASET_PATH, "metadata/non_sampleable_categories.txt")
+if os.path.isfile(non_sampleable_category_txt):
+    with open(non_sampleable_category_txt) as f:
+        NON_SAMPLEABLE_SYNSETS = set([FLOOR_SYNSET] + [line.strip() for line in f.readlines()])
+
 
 class OmniGibsonBDDLBackend(BDDLBackend):
     def get_predicate_class(self, predicate_name):
         return SUPPORTED_PREDICATES[predicate_name]
+
+
+class BDDLEntity(Wrapper):
+    """
+    Thin wrapper class that wraps an object or system if it exists, or nothing if it does not exist. Will
+    dynamically reference an object / system as they become real in the sim
+    """
+    def __init__(
+        self,
+        object_scope,
+        entity=None,
+    ):
+        """
+        Args:
+            object_scope (str): BDDL synset instance of the entity, e.g.: "almond.n.01_1"
+            entity (None or DatasetObject or BaseSystem): If specified, the BDDL entity to wrap. If not
+                specified, will initially wrap nothing, but may dynamically reference an actual object or system
+                if it exists in the future
+        """
+        # Store synset and other info, and pass entity internally
+        self.object_scope = object_scope
+        self.synset = "_".join(self.object_scope.split("_")[:-1])
+        self.is_system = self.synset in SUBSTANCE_SYNSET_MAPPING
+
+        # Infer the correct category to assign, special casing agents
+        if entity is not None and isinstance(entity, BaseRobot):
+            self.og_categories = ["agent"]
+        else:
+            self.og_categories = [SUBSTANCE_SYNSET_MAPPING[self.synset]] if self.is_system else \
+                OBJECT_TAXONOMY.get_subtree_igibson_categories(self.synset)
```

Comment:
maybe no special casing is needed if we add agent to be a valid synset?

Reply:
done!

Reply:
added todo to verify later